python/lsst/daf/butler/core/composites.py

```diff
@@ -0,0 +1,300 @@
+        name as derived from the `StorageClass`.
+    """
+    components = {}
+    if storageClass is not None and storageClass.components:
+        for c in storageClass.components:
+            if isinstance(composite, collections.Mapping):
+                comp = composite[c]
+            else:
+                try:
+                    comp = genericGetter(composite, c)
+                except AttributeError:
+                    pass
+                else:
+                    if comp is not None:
+                        components[c] = comp
+    return components
+
+
+def hasComponent(composite, componentName):
+    """Determine if it seems likely that the composite has the component.
+
+    Parameters
+    ----------
+    composite : `object`
+        Item to query for the component.
+    componentName : `str`
+        Name of component to retrieve.
+
+    Returns
+    -------
+    getter : `str`
+        Name of attribute that matched. None if nothing suitable
+        was found.
+
+    """
+    for attr in _attrNames(componentName, getter=True):
+        if hasattr(composite, attr):
+            return attr
+    return None
+
+
+def genericGetter(composite, componentName):
+    """Attempt to retrieve component from composite object by heuristic.
+
+    Will attempt a direct attribute retrieval, or else getter methods of the
+    form "get_componentName" and "getComponentName".
+
+    Parameters
+    ----------
+    composite : `object`
+        Item to query for the component.
+    componentName : `str`
+        Name of component to retrieve.
+
+    Returns
+    -------
+    component : `object`
+        Component extracted from composite.
+
+    Raises
+    ------
+    AttributeError
+        The attribute could not be read from the composite.
+    """
+    component = None
+    for attr in _attrNames(componentName, getter=True):
+        if hasattr(composite, attr):
+            component = getattr(composite, attr)
+            if attr != componentName:  # We have a method
+                component = component()
+            break
+    else:
+        raise AttributeError("Unable to get component {}".format(componentName))
+    return component
+
+
+def genericDisassembler(composite, storageClass, subset=None, override=None):
+    """Generic implementation of a disassembler.
+
+    This implementation attempts to extract components from the parent
+    by looking for attributes of the same name or getter methods derived
+    from the component name.
+
+    Parameters
+    ----------
+    composite : `object`
+        Parent composite object consisting of components to be extracted.
+    storageClass : `StorageClass`
+        `StorageClass` associated with the parent, with defined components.
+    subset : `iterable`, optional
+        Iterable containing subset of components to extract from composite.
+        Must be a subset of those defined in `storageClass`.
```

Comment:
I added this feature (and the override) when I started working in the real world of Exposure. The problem with Exposure is that some of the components are part of Exposure but others are one level down and part of `ExposureInfo`. Since I didn't have an ExposureInfo StorageClass (maybe I should have?) I couldn't tell it not to get upset when it couldn't get the calibs or whatever from the `Exposure`.